### PR TITLE
Refactor provisioner

### PR DESCRIPTION
**Pull request by samdoran**
Complete rewrite of the AWS lab provisioner to greatly improve efficiency. The entire playbook run now takes approximately one minute per instance.

Each component of the provisioning process is a separate role. Each role targets only the appropriate hosts, avoiding copious amounts of skipping.

Improve the README so it is much easier to follow.

@thisdavejohnson @tima @wtcross Please try this out and see if it resolves email and performance issues.

### CODE REVIEW

File: aws_lab_setup/roles/managed_node/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    - sudo
+
+- name: Install firewalld and libselinux-python
+  yum:
+    name:
+      - libselinux-python
+      - firewalld
+    state: present
+  when:
+    - "ansible_os_family == 'RedHat' or ansible_os_family == 'CentOS'"
```
Comment (tima) [inline aws_lab_setup/roles/managed_node/tasks/main.yml]: ansible_os_family will never be "CentOS." were you thinking of a different fact like ansible_distribution?

File: aws_lab_setup/roles/control_node/files/ansible.cfg [hunk 2]
```diff
@@ -0,0 +1,6 @@
+[defaults]
+connection = smart
+timeout = 60
+forks = 10
+inventory = ~/lightbulb/lessons/lab_inventory/
```
Comment (tima) [inline aws_lab_setup/roles/control_node/files/ansible.cfg]: inventory should be a variable that defaults to that path.

File: aws_lab_setup/roles/control_node/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+- name: Create lab inventory directory
+  file:
+    state: directory
+    path: /home/{{ username }}/lightbulb/lessons/lab_inventory
+
+- name: Put student inventory in proper spot
+  copy:
+    src: ./{{ username }}-instances.txt
+    dest: /home/{{ username }}/lightbulb/lessons/lab_inventory/hosts
```
Comment (tima) [inline aws_lab_setup/roles/control_node/tasks/main.yml]: Is this the best right place to put it for workshops/? what if we flatten the lightbulb examples/lessons/workshops?

File: aws_lab_setup/roles/managed_node/tasks/main.yml (hunk 1 above)
Comment (samdoran) [inline aws_lab_setup/roles/managed_node/tasks/main.yml]: Fixed in 68e8a6f9118079cbb0a8c3511a2e55c90f05dce4

File: aws_lab_setup/roles/control_node/files/ansible.cfg (hunk 2 above)
Comment (samdoran) [inline aws_lab_setup/roles/control_node/files/ansible.cfg]: Fixed in d42d391915fb2446cb5ee615c95c0091bc1ff15b

File: aws_lab_setup/roles/control_node/tasks/main.yml (hunk 3 above)
Comment (samdoran) [inline aws_lab_setup/roles/control_node/tasks/main.yml]: Probably not. But since it's now a variable it can be easily moved.

**Comment by tima** [comment]
@samdoran Ran into a roadblock testing this (as in not having time to fully resolve it) for now in that the embedded ec2_instance_types data is specific to the east region. I've been using the west region (also what's in the README) and hit errors because the AMIs cannot be found. 

In keeping with our goals for this, we need to create separate data structures for AMIs by region. Also thinking the ec2_lab_node_types maybe should be separate for the different types/configurations we may need to support.

Thinking we need an AMI mapping, instance type and a lab topology (lab_node_types) data. Need to think about how we feed that into the provision while giving other users quick and easy options for different labs types.

**Comment by tima** [comment]
@samdoran: what do you think about combining the common and user_accounts roles into one role? Not sure why we'd have one without the other.

**Comment by samdoran** [comment]
@tima `common` and `user_accounts` can be combined. Just how it shook out while I was making it. I was going for targeted roles rather than "junk drawer" roles. There's no technical reason they can't be combined. It's just organizational preference.

**Comment by samdoran** [comment]
The AMI thing is unfortunate. We could create a few dictionaries for each region, but this gets to be a pain to maintain. I would say no more than three regions just to keep maintenance down.

It should be structured such that specifying one of the three supported regions will pickup the correct AMIs. 

Are we thinking about letting folks customize the lab topology by adding or removing nodes? I hadn't thought about that.

**Comment by tima** [comment]
> It should be structured such that specifying one of the three supported regions will pickup the correct AMIs.

agreed. That's just what we package. The role should be setup that other users could provide their own.

> Are we thinking about letting folks customize the lab topology by adding or removing nodes? I hadn't thought about that.

Yeah that's where this exercise is heading -- many different uses from demos to informal training and from core to Tower to Windows to network automation. We'll have to shake out the details as we go, however; it won't always be 1 controller, 3 nodes and 1 proxy.

**Comment by tima** [comment]
I learned (a bit of the hard way with #70) that the provisioner fails when it tries to clone the lightbulb a second time. We need an `ignore_error: yes` on that task at a minimum.

**Comment by tima** [comment]
@samdoran do you have a sense when it would be good to merge this into the master? While there is a lot of work to be done (not just by you) to support the v2 effort, there is a lot of great work you've done here.

**Comment by samdoran** [comment]
@tima This is ready to be merged.

A review of the major changes:
- general overhaul of roles and structure
- add an option to control whether or not Ansible is installed
- support `es-west-1` and `us-east-1` regions with a framework for easily adding more regions
- updated AMI IDs to latest from the official owners
- add playbook to teardown as well as spin up lab
- add support for Ubuntu 14 and Ubuntu 16 (though 16 still needs a bootstrap task to work fully)
- use included SendGrid module that shipped with Ansible 2.2
- put `vimrc` in the proper place
- no longer create all user accounts on all systems
- vastly improve the time it takes to generate inventory files

There are still things that need to be done, like update the documentation, but this is stable enough to merge now. It's already getting a tad on the unwieldy side.

**Pull request merged**